Add bal dataloader to test SBT

hxu296 opened the pull request:
This PR adds a [BAL dataset](https://grail.cs.washington.edu/projects/bal/) loader at `pypose/examples/sparse/bal_loader.py`.

- The dataloader allows for reading Bundle Adjustment in the Large (BAL) datasets and is based on the implementation in [scipy's BAL loader in its non-linear least square demo](https://scipy-cookbook.readthedocs.io/items/bundle_adjustment.html). Scipy's results on BAL can also be used as a simple baseline to compare against pypose's bundle adjustment implementation.
- the dataloader accepts a `dataset` parameter, which is one of 'ladybug', 'trafalgar', 'dubrovnik', 'venice', or 'final'. Each BAL dataset corresponds multiple data files, and each data file is a standalone Bundle Adjustment problem.
- in each iteration, the dataloader will return a dictionary with the following fields:
  - problem_name: str The name of the problem.
  - camera_extrinsics: pp.SE3 (n_cameras, 7) The camera extrinsics, represented as pp.SE3. First three columns are translation, last four columns are unit quaternion.
  - camera_intrinsics: torch.Tensor (n_cameras, 3, 3) The camera intrinsics. Each camera is represented as a 3x3 K matrix.
  - points_3d: torch.Tensor (n_points, 3) Contains initial estimates of point coordinates in the world frame.
  - points_2d: torch.Tensor (n_observations, 2) Contains measured 2-D coordinates of points projected on images in each observation.
  - camera_indices: torch.Tensor (n_observations,) Contains indices of cameras (from 0 to n_cameras - 1) involved in each observation.
  - point_indices: torch.Tensor (n_observations,) Contains indices of points (from 0 to n_points - 1) involved in each observation.

hxu296 commented:
As discussed with Zitong, I will add a PyTorch SGD based BA algorithm to test the dataloader and also serve as a simple baseline for sparse LM based BA.

huyaoyu reviewed (changes requested):
Only reviewed the bal_loader.py. Need some more time on ba_dense.py.

hxu296 commented:
Thanks for your review @huyaoyu 
ba_dense.py is still WIP as for now. I will ping you when it's ready for review, thanks!

hxu296 commented on examples/sparse/bal_loader.py:
```diff
@@ -0,0 +1,179 @@
+
+import numpy as np
+from scipy.spatial.transform import Rotation
+from torchdata.datapipes.iter import HttpReader, IterableWrapper, OnlineReader
```
should I add a requirements.txt and README.md in `sparse` to document how to install dependencies and start the demo?

huyaoyu replied:
Yeah, that's a good idea. When we are more confident about the `sparse` module, we should then consider merging the requirements to pyposes' main requirement.txt file.

hxu296 commented on examples/sparse/bal_loader.py:
```diff
@@ -0,0 +1,179 @@
+            }
+
+def build_pipeline(dataset='ladybug', cache_dir='bal_data'):
+    assert dataset in ALL_DATASETS, f"dataset_name must be one of {ALL_DATASETS}"
```
sure!

hxu296 commented on examples/sparse/bal_loader.py:
```diff
@@ -0,0 +1,179 @@
+            contains initial estimates of point coordinates in the world frame.
+        - points_2d: torch.Tensor (n_observations, 2)
+            contains measured 2-D coordinates of points projected on images in each observations.
+        - camera_indices: torch.Tensor (n_observations,)
```
yeah, that's more clear I think

hxu296 commented on examples/sparse/bal_loader.py:
```diff
@@ -0,0 +1,179 @@
+def not_none(s):
+    return s is not None
+
+def debug(s):
```
There were a few unreferenced functions / imports for debug, I will trim them, thanks

hxu296 commented on examples/sparse/bal_loader.py:
```diff
@@ -0,0 +1,179 @@
+import numpy as np
+from scipy.spatial.transform import Rotation
+from torchdata.datapipes.iter import HttpReader, IterableWrapper, OnlineReader
+from torchvision.transforms import Compose
```
Yes

hxu296 commented on examples/sparse/bal_loader.py:
```diff
@@ -0,0 +1,179 @@
+    camera_intrinsics = np.zeros((n_cameras, 3, 3), dtype=np.float32)
+    camera_intrinsics[:, 0, 0] = camera_params[:, 6]
+    camera_intrinsics[:, 1, 1] = camera_params[:, 6]
+    camera_intrinsics[:, 0, 2] = camera_params[:, 7]
```
Thanks for catching that, and sorry for the confusion. I noticed that in the BAL camera model, the origin of the image is the center of the image. BAL also provides steps to project points to pixels. I noticed that pypose's native `point2pixel`and `reprojerr` functions did not account for radial distortions. I wonder where should I put my custome `bal_point2pixel`? I am thinking to put it in `ba_dense.py`, but the custom projection function is actually associated with the BAL dataset, but putting it in `ba_loader.py` feels wrong either. What do you think?

<img width="1430" alt="image" src="https://github.com/pypose/pypose/assets/61036578/d5de732c-ad48-4df2-869c-13a3948cca03">

hxu296 replied:
How about make a new `bal_utils.py` file and put a differentiable `ba_point2pixel` in there?

hxu296 replied:
done

huyaoyu replied:
OK, now I know the issues here. We need a whole new module representing camera-related projections, out of the sparse or ba moudles.

hxu296 commented:
The CI failure looks weird, as commit `535e907` is scoped within `examples/sparse`...

hxu296 commented on examples/sparse/bal_loader.py:
```diff
@@ -0,0 +1,179 @@
+
+    # convert Rodrigues vector to unit quaternion for camera rotation
+    # camera_params[0:3] is the Rodrigues vector
+    theta = np.linalg.norm(camera_params[:, :3], axis=1)[:, np.newaxis]
```
Hi, this part of code is removed as I just found out that scipy natively supports converting a Rodrigues vector to an unit quaternion with the following:

```
  r = Rotation.from_rotvec(camera_params[:, :3])
  q = r.as_quat()
```

Also, this seems to be the source of numerical instability when implementing the custom reprojerr function that considers radial distortions. Not sure why these lines are included in scipy's example.

hxu296 commented:
Addressed all comments regarding bal_loader.py. Added a new file bal_utils.py for a custom reprojerr function that considers radial distortions k1 and k2. I have verified that the custom reprojerr produces similar results as the native pp.reprojerr.

These two files are ready for review:
1. bal_loader.py
2. bal_utils.py

Meanwhile, I will keep working towards a bal_dense.py and a jupyter notebook example that is colab ready.

hxu296 commented:
Hi, the following two files are ready for review, and nothing else is on the workbench.

1. ba_dense.py
2. ba_example.ipynb

You can try the jupyter notebook example through google colab by clicking on the "Open in Colab" tab in the [README](https://github.com/hxu296/pypose/tree/sparse/ba/examples/sparse). 

### Here is a visualization of loss before and after optimization:

<img width="478" alt="image" src="https://github.com/pypose/pypose/assets/61036578/af25cb99-1562-4dba-925e-5d22ce47b625">


### Here is a visualization of the loss history:

<img width="668" alt="image" src="https://github.com/pypose/pypose/assets/61036578/51f92fcd-d752-4ea5-8d59-34e781059beb">

Cheers.

huyaoyu commented on examples/sparse/ba_dense.py:
```diff
@@ -0,0 +1,114 @@
+Link to the dataset: https://grail.cs.washington.edu/projects/bal/
+"""
+
+import torch, argparse, os
```
`os` is not used in this file.

hxu296 replied:
fixed. thanks for catching that!

huyaoyu commented on examples/sparse/ba_dense.py:
```diff
@@ -0,0 +1,114 @@
+    loss_history = []
+    loss_vector_history = []
+
+    for i in t:
```
`i` is not used in the for-loop body.

hxu296 replied:
replaced with `_` instead

huyaoyu commented on examples/sparse/ba_dense.py:
```diff
@@ -0,0 +1,114 @@
+    parser.add_argument('--steps', type=int, default=1000, help='Number of optimization steps.')
+    parser.add_argument('--dataset', type=str, default='ladybug', help='BAL dataset name.')
+    parser.add_argument('--problem', type=str, default='problem-49-7776-pre', help='BAL problem name.')
+    args = parser.parse_args()
```
I think we can add a `-h` or `--help` argument, which shows an example command line and direct the user to the `https://grail.cs.washington.edu/projects/bal/` webpage for possible values of the `--dataset` and `--problem` arguments.

hxu296 replied:
I refactored the help message into
```
usage: ba_dense.py [-h] [--steps STEPS] [--dataset DATASET] [--problem PROBLEM]

This script runs bundle adjustment on the BAL dataset. To understand what datasets and problems are available, please refer to https://grail.cs.washington.edu/projects/bal

options:
  -h, --help         show this help message and exit
  --steps STEPS      Number of optimization steps.
  --dataset DATASET  BAL dataset name, please refer to
                     https://grail.cs.washington.edu/projects/bal for possible
                     values.
  --problem PROBLEM  BAL problem name, please refer to
                     https://grail.cs.washington.edu/projects/bal for possible
                     values.

Example of optimizing `problem-49-7776-pre` in the `ladybug` dataset:
 python3 ba_dense.py --dataset ladybug --problem problem-49-7776-pre --steps 1000
```

huyaoyu commented on examples/sparse/bal_loader.py:
```diff
@@ -0,0 +1,254 @@
+    # camera_params[6] is focal length, camera_params[7] and camera_params[8] are two radial distortion parameters
+    # the origin of the image coordinate system is at the center of the image
+    camera_intrinsics = np.zeros((n_cameras, 3, 3), dtype=np.float32)
+    camera_intrinsics[:, 0, 0] = camera_params[:, 6]
```
Worth a comment saying that putting the distortion parameters in the intrinsic matrix is something like a hack and the customized projection error function will handle it correctly later.

huyaoyu commented on examples/sparse/bal_utils.py:
```diff
@@ -0,0 +1,160 @@
+
+def reprojerr(pose, points, pixels, intrinsics, distortions, point_index, camera_index):
+    """
+    Calculates batched per-pixel reprojection error considering camera distortion.
```
Give credit to the original author of the ba dataset.

huyaoyu commented on examples/sparse/bal_utils.py:
```diff
@@ -0,0 +1,160 @@
+    Returns:
+        Per-pixel reprojection error. The shape is (..., N).
+    """
+    # convert to camera coordinates
```
Let's have a quick meeting to discuss how to do proper comments to this function.

I have a hard time understanding the magic dimension numbers used in this function. I think future users might try to take a look at how projection errors are implemented and they might run into the same problem.

hxu296 replied:
After a discussion, Yaoyu and I reached the conclusion that we should integrate the image undistortion function into pypose's geometry module, and then simply call pypose native functions here. This will avoid exposing the complexity of raw dimension manipulations to the users and also strengthen pypose's API library.

A refactor is on the way.

huyaoyu reviewed (commented):
Some simple comments.

Need a quick call for discussion.

Thank you!

hxu296 commented:
Hello! Thanks for your review. I am up to a call after the monthly developer meeting through Slack or Zoom. My email is hxu296@wisc.edu.

hxu296 commented on examples/sparse/bal_loader.py:
```diff
@@ -0,0 +1,254 @@
+    """
+    global ALL_DATASETS
+    assert dataset in ALL_DATASETS, f"dataset_name must be one of {ALL_DATASETS}"
+    print(f"Streaming data for {dataset}...")
```
nice catch! fixed.
